File: backend/src/main/java/com/keresman/kanbanboard/service/AuthService.java
package com.keresman.kanbanboard.service;

import com.keresman.kanbanboard.model.User;
import com.keresman.kanbanboard.payload.LoginRequest;
import com.keresman.kanbanboard.payload.RegisterRequest;
import com.keresman.kanbanboard.payload.TokenResponse;

/**
 * Service interface defining authentication operations such as user registration, login, token
 * refresh, and logout.
 */
public interface AuthService {

  /**
   * Registers a new user and issues authentication tokens.
   *
   * @param registerRequest the request payload containing user registration details
   * @return a {@link TokenResponse} containing access and refresh tokens
   */
  TokenResponse register(RegisterRequest registerRequest);

  /**
   * Authenticates a user and issues authentication tokens.
   *
   * @param loginRequest the request payload containing login credentials
   * @return a {@link TokenResponse} containing access and refresh tokens
   */
  TokenResponse login(LoginRequest loginRequest);

  /**
   * Refreshes the access token using a valid refresh token.
   *
   * @param refreshTokenStr the refresh token string
   * @return a {@link TokenResponse} containing new access and refresh tokens
   */
  TokenResponse refresh(String refreshTokenStr);

  /**
   * Logs out a user by revoking their refresh tokens.
   *
   * @param user the user to log out
   */
  void logout(User user);
}

File: frontend/src/services/auth.service.ts
import { http, tokenStore } from "./http";
import type { LoginRequest, RegisterRequest, TokenResponse } from "../types";

export async function login(payload: LoginRequest): Promise<TokenResponse> {
    const { data } = await http.post<TokenResponse>("/api/v1/auth/login", payload);
    tokenStore.setTokens(data);
    return data;
}

export async function register(payload: RegisterRequest): Promise<TokenResponse> {
    const { data } = await http.post<TokenResponse>("/api/v1/auth/register", payload);
    tokenStore.setTokens(data);
    return data;
}

export async function refresh(refreshToken: string): Promise<TokenResponse> {
    const { data } = await http.post<TokenResponse>("/api/v1/auth/refresh", { refreshToken });
    tokenStore.setTokens(data);
    return data;
}

export async function logout(): Promise<void> {
    try {
        await http.post("/api/v1/auth/logout");
    } finally {
        tokenStore.clearTokens();
    }
}
